### PR TITLE
Azure site extension

### PR DESCRIPTION
**Pull request by Tratcher**
Here's the initial structure for the azure site extension. https://github.com/aspnet/AzureIntegration/issues/21.

### CODE REVIEW

**Comment by Tratcher** [comment]
Needs a friendly name
image

**Comment by Tratcher** [comment]
image

File: src/Microsoft.AspNetCore.AzureAppServices.SiteExtension/Microsoft.AspNetCore.AzureAppServices.SiteExtension.csproj [hunk 1]
```diff
@@ -0,0 +1,29 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <Import Project="..\..\build\common.props" />
+
+  <PropertyGroup>
+    <title>ASP.NET Core Extensions</title>
+    <Description>This extension enables additional functionality for ASP.NET Core on Azure WebSites, such as enabling Azure logging.</Description>
+    <TargetFramework>net451</TargetFramework>
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.AzureAppServices.SiteExtension/Microsoft.AspNetCore.AzureAppServices.SiteExtension.csproj]: Is this the right TFM? Or do we only need one to build this repo?

File: src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+            string result = null;
+            if (this.Arguments != null && this.Arguments.Count > 0)
+            {
+                foreach (string arg in this.Arguments)
+                {
+                    if (!string.IsNullOrWhiteSpace(arg))
+                    {
+                        string trimmedArg = arg.Trim();
+                        if (trimmedArg.ToUpperInvariant().StartsWith(name.ToUpperInvariant()))
```
Comment (pranavkm) [inline src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs]: `trimmedArg.StartsWith(name, StringComparison.OrdinalIgnoreCase);`

File: src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        protected override void Apply()
+        {
+            if (base.TargetChildNodes == null || base.TargetChildNodes.Count == 0)
```
Comment (pranavkm) [inline src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs]: Might not need `base.`

File: src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            {
+                base.TargetNode.AppendChild(base.TransformNode);
+            }
+            else
+            {
+                XmlAttribute transformAtt = null;
+
+                foreach (XmlAttribute att in this.TransformNode.Attributes)
+                {
+                    if (string.Compare(att.Name, this.AttributeName, StringComparison.InvariantCultureIgnoreCase) == 0)
```
Comment (pranavkm) [inline src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs]: `OrdinalIgnoreCase`

File: src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+                            {
+                                att.Value = transformAtt.Value;
+                            }
+                            else
+                            {
+                                //Horrible hack becuase this doesn't compose super nice with
+                                //insertOrAppend being applied on the TargetNode. The target node is created
+                                //with the children it has in the transform, which means we would duplicate the value
+                                //here.
+                                if (att.Value == transformAtt.Value) { return; }
```
Comment (pranavkm) [inline src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs]:
```
if (att.Value == transformAtt.Value)
{ 
  return; 
}
```
(formatting)

File: src/Microsoft.AspNetCore.AzureAppServices.SiteExtension/Microsoft.AspNetCore.AzureAppServices.SiteExtension.csproj (hunk 1 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.AzureAppServices.SiteExtension/Microsoft.AspNetCore.AzureAppServices.SiteExtension.csproj]: This matches the xdt dependency TFM

File: src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        public InsertOrAppendAttribute()
+            : base(TransformFlags.UseParentAsTargetNode, MissingTargetMessage.Error)
+        {
+        }
+
+        private string attributeName;
```
Comment (JunTaoLuo) [inline src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs]: underscore before private variables

File: src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+        /// <summary>
+        /// 
+        /// </summary>
+        protected string AttributeName
+        {
+            get
+            {
+                if (this.attributeName == null)
```
Comment (JunTaoLuo) [inline src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs]: no need for this

File: src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+                        if (trimmedArg.ToUpperInvariant().StartsWith(name.ToUpperInvariant()))
+                        {
+                            int start = arg.IndexOf('\'');
+                            int last = arg.LastIndexOf('\'');
+                            if (start <= 0 || last <= 0 || last <= 0)
+                            {
+                                throw new ArgumentException("Expected two ['] characters");
+                            }
+
+                            string value = trimmedArg.Substring(start, last - start);
```
Comment (JunTaoLuo) [inline src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs]: vars? is this file generated? It's quite different from our coding style.

File: src/Microsoft.Web.Xdt.Extensions/Microsoft.Web.Xdt.Extensions.csproj [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+  <PropertyGroup>
+    <Description>Additional functionality for Xdt transforms.</Description>
+    <TargetFramework>net451</TargetFramework>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
+    <PackageTags>xdt</PackageTags>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Web.Xdt" Version="1.4.0" />
```
Comment (JunTaoLuo) [inline src/Microsoft.Web.Xdt.Extensions/Microsoft.Web.Xdt.Extensions.csproj]: versions usually live in dependency.props no?

File: src/Microsoft.AspNetCore.AzureAppServices.SiteExtension/Microsoft.AspNetCore.AzureAppServices.SiteExtension.csproj [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  </PropertyGroup>
+
+  <ItemGroup>
+    <None Remove="applicationHost.xdt" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <Content Include="applicationHost.xdt" />
+    <Content Condition="$(Configuration)=='Debug'" Include="bin\Debug\net451\Microsoft.Web.Xdt.Extensions.dll" PackagePath="content" />
+    <Content Condition="$(Configuration)=='Release'" Include="bin\Release\net451\Microsoft.Web.Xdt.Extensions.dll" PackagePath="content" />
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.AzureAppServices.SiteExtension/Microsoft.AspNetCore.AzureAppServices.SiteExtension.csproj]: Include="bin\$(Configuration\...)"

File: src/Microsoft.AspNetCore.AzureAppServices.SiteExtension/Microsoft.AspNetCore.AzureAppServices.SiteExtension.csproj [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+    <Description>This extension enables additional functionality for ASP.NET Core on Azure WebSites, such as enabling Azure logging.</Description>
+    <TargetFramework>net451</TargetFramework>
+    <GenerateDocumentationFile>false</GenerateDocumentationFile>
+    <PackageTags>aspnet;logging;aspnetcore</PackageTags>
+    <IncludeBuildOutput>false</IncludeBuildOutput>
+    <ContentTargetFolders>content</ContentTargetFolders>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <None Remove="applicationHost.xdt" />
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.AzureAppServices.SiteExtension/Microsoft.AspNetCore.AzureAppServices.SiteExtension.csproj]: why do you remove applicationHost.xdt here and add it to content below?

**Comment by Tratcher** [comment]
FYI I'll only be applying packaging feedback for this pass. I'll revisit code feedback on a second pass, add tests, etc..

File: src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+namespace Microsoft.Web.Xdt.Extensions
+{
+    /// <summary>
+    /// Insert or append to the given attribute
+    /// </summary>
+    public class InsertOrAppendAttribute : Transform
+    {
+        /// <summary>
+        /// 
+        /// </summary>
```
Comment (anurse) [inline src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs]: Clean these up?

File: src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+        {
+            if (TargetChildNodes == null || TargetChildNodes.Count == 0)
+            {
+                TargetNode.AppendChild(TransformNode);
+            }
+            else
+            {
+                XmlAttribute transformAtt = null;
+
+                foreach (XmlAttribute att in TransformNode.Attributes)
```
Comment (anurse) [inline src/Microsoft.Web.Xdt.Extensions/InsertOrAppendAttribute.cs]: Why not use `.FirstOrDefault`? If it's a performance thing, comment it as such.

File: src/Microsoft.AspNetCore.AzureAppServices.SiteExtension/applicationHost.xdt [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+<configuration xmlns:xdt="http://schemas.microsoft.com/XML-Document-Transform">
+
+  <xdt:Import path="%XDT_EXTENSIONPATH%\Microsoft.Web.Xdt.Extensions.dll"
+              namespace="Microsoft.Web.Xdt.Extensions" />
+
+    <system.webServer xdt:Transform="InsertIfMissing">
+      <runtime xdt:Transform="InsertIfMissing" >
+        <environmentVariables xdt:Transform="InsertIfMissing">
+          <add name="DOTNET_ADDITIONAL_DEPS" value="%XDT_EXTENSIONPATH%\deps" xdt:Locator="Match(name)" xdt:Transform="InsertOrAppendAttribute(Attribute='value')" />
+          <add name="ASPNETCORE_HOSTINGSTARTUPASSEMBLIES" value="Microsoft.AspNetCore.AzureAppServices.HostingStartup" xdt:Locator="Match(name)" xdt:Transform="InsertOrAppendAttribute(Attribute='value')" />
```
Comment (anurse) [inline src/Microsoft.AspNetCore.AzureAppServices.SiteExtension/applicationHost.xdt]: Is this the normal pattern for XDT custom transforms? The key-value thing seems to require a lot of extra work. Couldn't it just be `InsertOrAppendAttribute(value)`?

**Pull request merged**